src/Compilers/CSharp/Portable/Symbols/TypeSymbolExtensions.cs

```diff
@@ -525,13 +525,67 @@ public static SpecialType GetSpecialTypeSafe(this TypeSymbol? type)
 
         public static bool IsAtLeastAsVisibleAs(this TypeSymbol type, Symbol sym, ref HashSet<DiagnosticInfo>? useSiteDiagnostics)
         {
-            HashSet<DiagnosticInfo>? localUseSiteDiagnostics = useSiteDiagnostics;
-            var result = type.VisitType((type1, symbol, unused) => IsTypeLessVisibleThan(type1, symbol, ref localUseSiteDiagnostics), sym,
+            var closure = IsAtLeastAsVisibleAsClosure.Get(useSiteDiagnostics);
+
+            var result = type.VisitType(closure.Func, sym,
                                         canDigThroughNullable: true); // System.Nullable is public
-            useSiteDiagnostics = localUseSiteDiagnostics;
+
+            useSiteDiagnostics = closure.Release();
             return result is null;
         }
 
+        // Type to reduce allocations of the Func<TypeSymbol, Symbol, bool, bool> closure in the above method
+        private class IsAtLeastAsVisibleAsClosure
+        {
+            [ThreadStatic]
+            private static IsAtLeastAsVisibleAsClosure? s_cache;
+
+            private HashSet<DiagnosticInfo>? _useSiteDiagnostics;
+            public Func<TypeSymbol, Symbol, bool, bool> Func { get; }
+
+            private IsAtLeastAsVisibleAsClosure()
+            {
+                Func = IsTypeLessVisibleThan;
+            }
+
+            private bool IsTypeLessVisibleThan(TypeSymbol type, Symbol symbol, bool unused)
+                => TypeSymbolExtensions.IsTypeLessVisibleThan(type, symbol, ref _useSiteDiagnostics);
+
+            public static IsAtLeastAsVisibleAsClosure Get(HashSet<DiagnosticInfo>? useSiteDiagnostics)
+            {
+                IsAtLeastAsVisibleAsClosure value;
+
+                ref var cache = ref s_cache;
+                if (cache is not null)
+                {
+                    value = cache;
+                    cache = null;
+                }
+                else
+                {
+                    value = new IsAtLeastAsVisibleAsClosure();
+                }
+
+                value._useSiteDiagnostics = useSiteDiagnostics;
+
+                return value;
+            }
+
+            public HashSet<DiagnosticInfo>? Release()
+            {
+                var useSiteDiagnostics = _useSiteDiagnostics;
+                _useSiteDiagnostics = null;
+
+                ref var cache = ref s_cache;
+                if (cache is null)
+                {
+                    cache = this;
+                }
+
+                return useSiteDiagnostics;
+            }
+        }
```

Comment:
Tbh this seems really complex and hard to validate. It also seems like a lot of bloat if we wanted to use this in other locations. Would it be possible to somehow encapsulate this into some generic class we could review once and then use when appropriate?

In other words, I really dislike all this one-off patching of perf issues with non reusable solutions. The cost/benefit ratio just isn't there for me in terms of maintenance costs.